Updated GpuInterop sample (fixed calculating yaw, pitch and row; vertically flipped Vulkan image)

abenedik opened the pull request:
The first commit in the PR updates the sample so that when yaw, pitch and roll are zero the model is viewed straight into the front side. Commit also fixes calculating worldViewProj in D3D11DemoControl (before changing the Yaw slider changed pitch angle).

In the second commit in the PR, the rendered Vulkan image is vertically flipped so that on the rendered 2D image the y axis will point up. To flip the image the Height in Viewport is set to negative height value and Y value is offset by height value. See more: https://www.saschawillems.de/blog/2019/03/29/flipping-the-vulkan-viewport/


## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->


## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->


## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->


## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->


## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
<!--- List any breaking changes here. -->

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

abenedik committed:
Fixed using YawPitchRoll in GpuInterop sample
This updates the sample so that when yaw, pitch and roll are zero the model is viewed straight into the front side.
Also fixes calculating worldViewProj in D3D11DemoControl.

abenedik committed:
Vertically flip the rendered 2D image so y axis points up

kekekeks commented:
The Y flip is supposed to be controlled via `PlatformGraphicsExternalImageProperties.TopLeftOrigin` property

abenedik commented:
> The Y flip is supposed to be controlled via `PlatformGraphicsExternalImageProperties.TopLeftOrigin` property

Ok, thank you, I missed that property. When it is set to true, then the image is correctly shown.

But I must say that the property name and its meaning are not intuitive.

To improve that it would be good that this property is set to true by default because users usually do not want to have the image vertically flipped.

Even better would be if that property would be renamed into a property that can have its default value set to false - for example the name of the property could be VerticallyFlipImage. In this case its default value can be false and also user would know exactly what to expect from that property - the TopLeftOrigin as Boolean is not very clear (especially what the default false value means - the origin is not in the top-left, where is it then?).

kekekeks commented:
I believe that with DirectX rendering the image won't be flipped.

abenedik commented:
The current GpuInterop sample shows a vertically flipped image for Vulkan and DirectX interop. Apply the changes in the first commit in this PR (to fix using yaw, pitch and roll) and then see that the bottom plane (yellow color) is shown on top (see also comments in the email that I have sent you).

abenedik commented:
The naming in Skia is better because the origin has two well defined enum values: `TopLeft` and `BottomLeft`. Also, the `TopLeft` is the default option.

abenedik commented:
Is it possible to enable Pointer events (PointerMoved, PointerPressed, etc.) on controdo not work anymore on the Control that shows the 3D scene (D3D11DemoControl, VulkanDemoControl).

abenedik commented:
Is it possible to enable Pointer events (PointerPressed, PointerMoved etc.) on the Control that shows the 3D scene (D3D11DemoControl, VulkanDemoControl).

Pointer events worked when using the previous interop method with shared OpenGL context (preview 4), but now with the new interop they do not work anymore. Is there a setting that need to be enabled to make this work.

abenedik committed:
Prevented memory leak in D3D11SwapchainImage when resizing
Updated DisposeAsync method by adding disposing ICompositionImportedGpuImage object. This is was already done in VulkanSwapchainImage.DisposeAsync.

abenedik commented:
I have added another commit to this PR - this prevents a memory leak when using `D3D11SwapchainImage` in GpuInterop project.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0031523-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

jmacato commented:
Closing this PR temporarily due to inactivity. Please ping me if this needs to be reopened.
